File: kotlin/canConstruct.kt
fun canConstruct(target: String, wordBank: Array<String>): Boolean {
    if (target.isEmpty()) return true;

    wordBank.forEach { word ->
        if (target.startsWith(word)) {
            val suffix: String = target.slice(word.length..(target.length - 1))
            if (canConstruct(suffix, wordBank)) return true
        }
    }

    return false
}

fun main(): Unit {
    println(canConstruct("skateboard", arrayOf("bo", "rd", "ate", "t", "ska", "sk", "boar")))
}

File: javascript/canConstruct.js
function canConstruct(target, wordBank, memo = {}) {
    if (target.length === 0) return true
    if (target in memo) return memo[target]
    
    for (let word of wordBank) {
        if (target.startsWith(word)) {
            const suffix = target.substr(word.length)
            memo[target] = canConstruct(suffix, wordBank, memo)

            if (memo[target]) return true
        }
    }

    memo[target] = false
    return false
}

console.log(canConstruct("skateboard", ["bo", "rd", "ate", "t", "ska", "sk", "boar"]))
console.log(canConstruct("eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeef",
["e", "ee", "eee", "eeee", "eeeee", "eeeeee"]))